lib/Sema/TypeChecker.cpp

```diff
@@ -375,6 +375,44 @@ void swift::performWholeModuleTypeChecking(SourceFile &SF) {
   }
 }
 
+void swift::loadDerivativeConfigurations(SourceFile &SF) {
+  if (!isDifferentiableProgrammingEnabled(SF))
```

Comment:
When is this true? The recursive walk over the AST in secondary files is going to trigger delayed parsing in all nominal types and extensions, which will impact performance in non-WMO builds.

Reply:
`isDifferentiableProgrammingEnabled` returns true only for files that have `import _Differentiation`, so I hope there would be negligible impact for other files.

Reply:
@slavapestov This is true when there is an `import _Differentiation` in the given SourceFile (so, corresponding ImportDecl).

What is the viable alternative? Essentially the task is: for function `f` find all functions that has `@derivative(of : f)` attribute specified. Both in primary and secondary sources.

Reply:
> What is the viable alternative? Essentially the task is: for function f find all functions that has @derivative(of : f) attribute specified. Both in primary and secondary sources.

The only real alternative would be to change the design of this attribute so that this kind of search does not have to be performed. By definition, it requires parsing all source files ahead of time which is incompatible with our strategy of skipping bodies of nominal types and extensions in secondary files.

Reply:
Yeah. Though the alternative design would not easily allow "pluggable" derivatives.... (see https://github.com/apple/swift/pull/58644#issuecomment-1119158991 for the design explanation, etc. by @dan-zheng )

Reply:
One alternative is to handle this like we do operators and AnyObject lookup. When skipping over the body of a nominal type or extension, make a note if the lexer sees '@_derivative', then eagerly parse only those delayed bodies when performing this global lookup. That's still not great but better than disabling delayed parsing entirely when 'import _Differentiation' appears in the program.

Reply:
@slavapestov Yes, I though about this as well, thanks! Though I was a bit hesitant to introduce such derivative-specific logic into lexer and parser :) But if you're thinking it's acceptable, I can try to do something around this in the subsequent PRs.